### PR TITLE
Final Capstone Project - Book an Appointment

### PR DESCRIPTION
**Pull request by yigitm**
In this PR:

- rails API is created with Postgres database
- Add/Delete/List endpoints are created.
- Basic authentication is created with JWT & bcrpyt
- API documentation is created with rswag
- Documentation configuration is done with swagger YAML
- Styling & design is based on the design mentioned in instructions 

[Project Requirements](https://github.com/microverseinc/curriculum-final-capstone/blob/main/projects/business_requirements.md)

**After running 'heroku run rails db:migrate' command; I got the below error. That's why for now on; I can only provide the front end's live version link. 

heroku-error

**If you run the Rails API in your localhost/3000; you can test the [Live version of Front End](https://nimble-hummingbird-7623a3.netlify.app/).

[Latest Branch For Front End Repository](https://github.com/yigitm/book-appointment-frontend/tree/Readme-Frontend)  

** I would be grateful if you may have any suggestions to fix this error. Thank you for your time.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,24 +1,182 @@
...
+rails db:migrate
+rails db:seed
+```
+
+To fetch the latest version of the app, run below commands.
+
+```
+git checkout -b API-readme
+git pull origin API-readme
+```
```
Comment (mcrd25) [inline README.md]: this should probably come after cd command

Comment (mcrd25) [reply README.md]: This is also very specific to this branch - it'd not be case for someone who is cloning repo for use - probably would be dev or main branch so don't forget to change this after merge

File: README.md [hunk 2]
```diff
@@ -1,24 +1,182 @@
...
+bundle install
+```
+
+Run below to set database.
+
+```
+rails db:create
+rails db:migrate
+rails db:seed
```
Comment (mcrd25) [inline README.md]: When I run this, I get following error:
image

Comment (mcrd25) [reply README.md]: check how the seeds are created - see if maybe you are missing something in instructions before seeding

File: README.md (hunk 1 above)
Comment (mcrd25) [reply README.md]: You are also missing instructions on running tests

File: spec/models/course_spec.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  end
+
+  it 'course info should be present' do
+    subject.info = nil
+    expect(subject).to_not be_valid
+  end
+
+  it 'course type length should be >= 10' do
+    subject.info = 'Pc'
+    expect(subject).to_not be_valid
```
Comment (mcrd25) [inline spec/models/course_spec.rb:31]: It's great that you've done tests but all I see are negative tests - when you expect it to fail. What about when you expect them to be successful. You'd be surprised how many times in the real world something you expect to work does not/ Doing positive tests is also important. It'd save you time while developing and help you catch bugs before they get too messy.

Comment (mcrd25) [reply spec/models/course_spec.rb:31]: Do above for all tests ⬆️

File: spec/models/detail_spec.rb [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    subject.course_name = nil
+    expect(subject).to_not be_valid
+  end
+
+  it 'course name length should be >= 3' do
+    subject.course_name = 'A3'
+    expect(subject).to_not be_valid
+  end
+
+  it 'course summary should be present' do
```
Comment (mcrd25) [inline spec/models/detail_spec.rb:18]: See comments on positive tests above

File: spec/models/user_spec.rb [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    expect(subject).to_not be_valid
+  end
+
+  it 'name should be <= 30 letters' do
+    subject.username = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcde'
+    expect(subject).to_not be_valid
+  end
+
+  it 'username should be unique & not be case sensitive' do
+    subject_two = User.new(username: 'test user', password: '123456')
```
Comment (mcrd25) [inline spec/models/user_spec.rb:24]: See comments on positive tests above

File: spec/requests/api/v1/courses_spec.rb [hunk 6]
```diff
@@ -0,0 +1,5 @@
+require 'swagger_helper'
+# rubocop:disable Lint/EmptyBlock
+RSpec.describe 'api/v1/courses', type: :request do
+end
```
Comment (mcrd25) [inline spec/requests/api/v1/courses_spec.rb:4]: Please add tests for this section. Tests ought to be comprehensive

File: app/controllers/api/v1/courses_controller.rb [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+      before_action :set_course, only: %i[show update]
+
+      # GET /courses
+      def index
+        @courses = Course.all
+
+        if user_authorized?
+          render json: @courses
+        else
+          render json: { error: 'Invalid API token! User not found.' }, status: :unprocessable_entity
```
Comment (mcrd25) [inline app/controllers/api/v1/courses_controller.rb:15]: I think this maybe can be a method. Since it's something you'd continuously do. You can pass type of message (error, success etc., message itself etc.) It'd make for cleaner code and won't have to repeat self as much

**Review by mcrd25 — CHANGES_REQUESTED** [review]
Hi @yigitm, 

Good job so far!
There are some issues that you still need to work on to prepare your project for the final evaluation but you are almost there!



## Suggested changes
_Check the comments under the review._

You can use as many of my suggestions as you want. If there is anything you would like to skip - feel free to do that. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- Writing more tests (ensure negative and positive scenarios are covered. You are currently missing tests for requests)
- Ensure you don't repeat yourself in code too much - If that's the case then you can create a method that does same and use that
- I may have missed some things so ensure that you are following good coding practices and improve code where necessary.
- Concerning your issue on heroku: Your problem lies in your Gemfile, see if yo ucan figure it out by reading this thread: https://github.com/rswag/rswag/issues/261

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

File: README.md (hunk 1 above)
Comment (yigitm) [reply README.md]: Thank you. I will definitely do that. By the way, thank you so much again, I changed my Gem file and it solved my deploy issue.

File: README.md (hunk 2 above)
Comment (yigitm) [reply README.md]: Thank you. I totally missed that, I was using these seeds at beginning of the project. I removed them.

File: spec/requests/api/v1/courses_spec.rb (hunk 6 above)
Comment (yigitm) [reply spec/requests/api/v1/courses_spec.rb:4]: Thank you. I will try to add more comprehensive tests.

File: app/controllers/api/v1/courses_controller.rb (hunk 7 above)
Comment (yigitm) [reply app/controllers/api/v1/courses_controller.rb:15]: thank you. I did refactor in the front end as pointed out by you but I got limited time to project presentation today, but I will allocate some time to refactor near future.

File: spec/models/course_spec.rb (hunk 3 above)
Comment (yigitm) [reply spec/models/course_spec.rb:31]: Thanks for that. Not enough but I added some positive tests to all models.

File: README.md (hunk 1 above)
Comment (mcrd25) [reply README.md]: That's no problem.
If you liked my review, feel free to connect on LinkedIn and leave a skill endorsement: https://www.linkedin.com/in/mayadouglas/

You are also free to follow me on GitHub
